crates/proof-impl/eth-ee-acct/src/lib.rs

```diff
@@ -0,0 +1,170 @@
+    // Each block is: [exec_block_package (SSZ)][raw_block_body (strata_codec)]
+    let num_blocks: u32 = zkvm.read_borsh();
+    let mut serialized_blocks = Vec::with_capacity(num_blocks as usize);
+    for _ in 0..num_blocks {
+        serialized_blocks.push(zkvm.read_buf());
+    }
+
+    // Build CommitChainSegment from blocks
+    let commit_segments = build_commit_segments_from_blocks(serialized_blocks)
+        .expect("Failed to build commit segments from blocks");
+
+    // Already raw bytes
+    let raw_prev_header = zkvm.read_buf();
+    let raw_partial_pre_state = zkvm.read_buf();
+
+    // Read genesis data for ChainSpec construction
+    let genesis: rsp_primitives::genesis::Genesis = zkvm.read_serde();
+    let chain_spec: Arc<ChainSpec> = Arc::new((&genesis).try_into().expect("Invalid genesis"));
+    let ee = EvmExecutionEnvironment::new(chain_spec);
+
+    let shared_private =
+        SharedPrivateInput::new(commit_segments, raw_prev_header, raw_partial_pre_state);
+
+    // Verify and apply the update operation
+    // This internally verifies the extra_data fields (processed_inputs, etc.)
+    verify_and_apply_update_operation(
+        &mut astate,
+        &operation,
+        coinputs.iter().map(|v| v.as_slice()),
+        &shared_private,
+        &ee,
+    )
+    .expect("Update verification failed");
+
+    // The operation already contains the correct final ProofState
+    let final_proof_state = operation.new_state();
+
+    // Extract UpdateExtraData from the operation (already verified by runtime)
+    let extra_data = decode_buf_exact::<UpdateExtraData>(operation.extra_data())
+        .expect("Failed to decode UpdateExtraData");
+
+    // Build the complete public output
+    let proof_output = EthEeAcctProofOutput {
+        prev_state: prev_proof_state,
+        final_state: final_proof_state,
+        da_commitments: Vec::new(), // Empty for now, TODO: when do we actually fill it??
+        output_messages: operation.outputs().messages().to_vec(),
+        input_messages: operation.processed_messages().to_vec(),
+        extra_data,
+    };
+
+    // Commit the complete output
+    zkvm.commit_borsh(&proof_output);
+}
+
+/// Verify that the astate hash matches the expected ProofState
+///
+/// This ensures that the initial account state matches what the proof claims as the starting point.
+fn verify_proof_state_matches(astate: &EeAccountState, expected_proof_state: &ProofState) {
+    // Compute tree_hash_root of the astate
+    let computed_hash = TreeHash::<Sha256Hasher>::tree_hash_root(astate);
+    let computed_hash = strata_identifiers::Hash::from(computed_hash.0);
+
+    // Compare with the expected inner_state from ProofState
+    if computed_hash != expected_proof_state.inner_state() {
+        panic!(
+            "Account state hash mismatch: computed {:?}, expected {:?}",
+            computed_hash,
+            expected_proof_state.inner_state()
+        );
+    }
```

Comment:
Should use `assert_eq!` here, and can probably avoid string interpolation.  This only ever uses `expected_proof_State.inner_state()` so we should just pass the inner state instead of the whole `ProofState`.